doc/sphinx/make-your-own-compositor.md

```diff
@@ -0,0 +1,253 @@
+```cmake
+cmake_minimum_required(VERSION 3.5)
+
+project(demo-mir-compositor)
+
+set(CMAKE_CXX_STANDARD 20)
+
+include(FindPkgConfig)
+pkg_check_modules(MIRAL miral REQUIRED)
+pkg_check_modules(XKBCOMMON xkbcommon REQUIRED)
+
+add_executable(demo-mir-compositor main.cpp)
+
+target_include_directories(demo-mir-compositor PUBLIC SYSTEM ${MIRAL_INCLUDE_DIRS})
+target_link_libraries(     demo-mir-compositor               ${MIRAL_LDFLAGS})
+target_link_libraries(     demo-mir-compositor               ${XKBCOMMON_LIBRARIES})
+```
+
+Then to build:
+```sh
+cmake -B build
+cmake --build build
+```
+
+To run, you can run nested in an X or Wayland session, or from a virtual
+terminal, just like the demo applications in [Explore What Mir Is Capable of
+Using Mir Demos](explore-mir-using-demos.md). For example, if we were to run
+inside an existing Wayland session, we'd run the following command:
+```sh
+WAYLAND_DISPLAY=wayland-99 ./build/demo-mir-compositor 
+```
+You'll see a window housing the compositor with a black void filling it. To
+fill this void with some content, you can run the following from another
+terminal:
+```sh
+WAYLAND_DISPLAY=wayland-99 bomber
+```
+You're free to replace `bomber` with any other Wayland compatible application.
+
+### Startup Applications
+When starting a compositor or display server, you usually want to start a few
+other applications. For example you might want to start a background viewer, a
+terminal, and a status bar. Doing so manually like what we've do would be
+immensely annoying, so let's add an option for users to pass a list of startup
+apps!
+
+It's up to you how you want your command line interface will be. Let's say
+we'll take a list of colon separated program names, then we'll launch each of
+those individually. 
+
+We'll start by creating an instance of `miral::ExternalClientLauncher`, this
+will provide a way to launch programs by name. Then, we'll create a function
+that will take the colon separated string and launch each application
+separately. Finally, we just need to pass the external client launcher and our
+configuration option to {func}`run_with` so that they're registered with the
+compositor.
+
+```diff
+ #include <miral/runner.h>
++#include <miral/configuration_option.h>
++#include <miral/external_client.h>
+ #include <miral/minimal_window_manager.h>
+ #include <miral/set_window_management_policy.h>
+ 
+@@ -9,8 +10,22 @@ int main(int argc, char const* argv[])
+ {
+     MirRunner runner{argc, argv};
+ 
++    miral::ExternalClientLauncher external_client_launcher;
++    
++    auto run_startup_apps = [&](std::string const& apps) 
++    {
++        for(auto i = std::begin(apps); i != std::end(apps);)
++        {
++            auto const j = std::find(i, std::end(apps), ':');
++            external_client_launcher.launch(std::vector<std::string>{std::string{i, j}});
++            if((i = j) != std::end(apps)) ++i;
++        }
++    };
```

Comment:
I know this comes from existing example code, but we can now add an option multiple times. (Maybe the answer could be updated.)

```suggestion
+    auto run_startup_apps = [&](std::vector<std::string> const& apps) 
+    {
+        for(auto const& app : apps)
+        {
+            external_client_launcher.launch(std::vector<std::string>{app}});
+        }
+    };
```

Reply:
One thing I just noticed while trying this out in code is that the compiler seems to not be able to infer that the lambda is convertible to `std::function<void(std::vector<std::string> const&)>`. If I replace `auto` with the type, it compiles.

Reply:
Obviously, examples should compile. (We should find a way to automate that.)

I thought the `lambda_as_function<>` hack provided the necessary level of indirection. What was the error message?

Reply:
@tarek-y-ismail see #3507